### ISSUE TITLE
Mirror Go

### ISSUE THREAD

**Issue by pheasant75** [gcp/leela-zero#894]
I am wondering how does the existing LZ handles Mirror Go.  Especially by white.  Has anyone tested it?
In addition, how do LZ learn to deal with this strategy in the future.  I guess none of the training game so far would "happen" to produce such play.  Shall we specifically train LZ on this scenario? It would be very easy to write a routine to mirror the play (even we can set threshold of plays to stop mirroring).  But does this constitute "human knowledge"?

**Comment by gcp** [comment]
>But does this constitute "human knowledge"?

Yes.

**Comment by pheasant75** [comment]
Well, we didn't teach him how to deal with it though.

**Comment by jonlave** [comment]
I think the better question is, could LZ eventually learn to handle mirror go.

**Comment by wctgit** [comment]
Probably, IMHO. All that's really required is for it to recognize the asymmetry of tengen in the moves leading up to playing it. But perhaps an over-aggressive FPU algorithm will not allow it to explore symmetrical moves very much, because most of the time they will not be anywhere close to 'the best' move possible. So, it may take a carefully considered and balanced FPU algorithm to let it explore enough to try mirrored moves enough to learn from them. So, maybe not the current version, which is a bit of a jerry-rigged FPU calculation, and in a 'perfect' version it still might take quite a while before it plays enough self-play games to discover the importance of tengen and komi. In other words, it's probably a longer-term learning task, compared to the life-and-death stuff it seems to be learning now.

**Comment by pheasant75** [comment]
Yes, I guess it can learn to deal with it because that is what alphago_zero had achieved.
I suppose LZ will not afraid of white mirror play because:
White's strategy to use mirror go is to mirror if black plays the best move, and stop mirroring if there are better play than what black had played.  But since LZ has always learnt to play the best move, the human player who adopted mirror go strategy will never find opportunity to break from mirror early.  So the game went until to the point LZ find 10x10(TenGen) is the best next move when white can't mirror anymore.

**Comment by Eddh** [comment]
Where was it mentioned that alphago zero could deal with mirror go ?

**Comment by tterava** [comment]
The whole concept of mirror-go relies on being able to stay relatively even until you think you see a weakness in your opponent's play. That's when you break out of mirroring and do your own thing. If you try mirroring LZ's moves, you will quickly realize it's very difficult to find such weaknesses if you are a weaker player than LZ because LZ just approaches it like any other game and tries to find the strongest moves possible. Then at some point LZ will break the symmetry by playing Tengen and you realize you're losing.

Mirroring is fundamentally a flawed strategy and I think any attempt to counter it would just make LZ weaker.

**Comment by sethtroisi** [comment]
AlphaGo Master knows how to handle Mirror Go see https://alphagoteach.deepmind.com
Play out a mirror go sequence and you'll see that winrate for mirroring player drops.

This doesn't prove that Zero knows how to handle but it's a good test.
Similarly we could make a tool to test if Leelaz think it handles mirror go by looking at winrate after mirroring several moves in the opening.

**Comment by jkiliani** [comment]
I think that in general, a player mirroring with black should not pose a problem to Leela Zero. For the case of someone mirroring with white, it may be if LZ waits too long before playing the center point. A while ago, someone tried this successfully and posted on Reddit: https://www.reddit.com/r/cbaduk/comments/7uog1d/mirror_go_is_an_effective_strategy_for_us_weak/
In this game, white broke the mirror at move 174 and ended up winning.

**Comment by pheasant75** [comment]
@sethtroisi I play with it.  Black plays Q16, then Q3, Then use the recommended Q6 while white mirrors all the move.  The best play sequence after that is white mirroring and the winrate is 45.4.  Unfortunately the database end at step 23, so we can't see how black would deal with the mirror.

**Comment by Marcin1960** [comment]
What happens if black starts with K10 and mirrors after that?

**Comment by pheasant75** [comment]
@Marcin1960 black would be losing (TenGen) is a bad point (41.5%).  If black continue to mirror, the winrate will go down further.  White's move will not favor the play where black stone at TenGen would cast influence.

**Comment by pheasant75** [comment]
Just mean black mirror in general is a bad move.

**Comment by Marcin1960** [comment]
pheasant75

You did not understand my question.

**Comment by wctgit** [comment]
Black will be at a big disadvantage because of white's komi. The tengen point is not enough of an advantage to overcome komi by itself.

**Comment by Marcin1960** [comment]
wctgit

Thank you.

**Comment by tterava** [comment]
The reason Leela lost against the mirroring strategy in the game above is because it made a mistake in the capturing race. When LZ gets stronger, it won't do simple mistakes like that anymore.

AlphaGo isn't specifically trained to fight mirroring strategy, instead it intuitively knows that mirroring is a flawed strategy as I explained above and that's why it gives the mirroring moves lower winrate. Just play against LZ and see for yourself, it always ends up in situation where after tengen is played you're losing.

**Comment by wctgit** [comment]
Hmm, but actually setting up some sort of side experiment where we put LZ against a mirroring bot (which perhaps takes over with regular LZ at a certain point) might be interesting just to see where LZ currently stands against mirroring. Just for fun and info.

**Comment by Eddh** [comment]
I've actually seen recent games from dan players who are trying mirroring and LZ seems to do alright.

**Comment by codesinthedark** [comment]
Did someone try to change the code of LZ so that it mirror with white if it has more than 50% chance of winning, otherwise play something else. I would like to know how would black handle playing with the same engine that plays mirror go to a certain point? When would it play tengen, or when would it try to set double ladders or something similar. Does it have a chance of winning?
